x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/accesscontrol/IndicesPermissionTests.java

```diff
@@ -663,6 +670,95 @@ public void testIndicesPermissionHasFieldOrDocumentLevelSecurity() {
+            aliasMetadata,
+            List.of(
+                IndexMetadata.builder("logs-index")
+                    .settings(
+                        Settings.builder()
+                            .put(SETTING_NUMBER_OF_SHARDS, 1)
+                            .put(SETTING_NUMBER_OF_REPLICAS, 0)
+                            .put(SETTING_INDEX_VERSION_CREATED.getKey(), Version.CURRENT)
+                    )
+                    .putAlias(aliasMetadata)
+                    .build()
+            )
+        );
+        IndicesPermission.IsResourceAuthorizedPredicate predicate = new IndicesPermission.IsResourceAuthorizedPredicate(
+            StringMatcher.of("other"),
+            StringMatcher.of(dataStreamName, backingIndex.getName(), concreteIndex.getName(), alias.getName())
+        );
+        assertThat(predicate.test(dataStream), is(false));
+        // test authorization for a missing resource with the datastream's name
+        assertThat(predicate.test(dataStream.getName(), null), is(true));
+        assertThat(predicate.test(backingIndex), is(false));
+        // test authorization for a missing resource with the backing index's name
+        assertThat(predicate.test(backingIndex.getName(), null), is(true));
+        assertThat(predicate.test(concreteIndex), is(true));
+        assertThat(predicate.test(alias), is(true));
+    }
+
+    public void testResourceAuthorizedPredicateAnd() {
+        IndicesPermission.IsResourceAuthorizedPredicate predicate1 = new IndicesPermission.IsResourceAuthorizedPredicate(
+            StringMatcher.of("other", "a"),
+            StringMatcher.of("b", "x")
+        );
+        IndicesPermission.IsResourceAuthorizedPredicate predicate2 = new IndicesPermission.IsResourceAuthorizedPredicate(
+            StringMatcher.of("other", "b"),
+            StringMatcher.of("a", "x")
+        );
+        Metadata.Builder mb = Metadata.builder(
+            DataStreamTestHelper.getClusterStateWithDataStreams(
+                List.of(Tuple.tuple("a", 1)),
+                List.of(),
+                Instant.now().toEpochMilli(),
+                builder().build(),
+                1
+            ).getMetadata()
+        );
+        IndexAbstraction.DataStream dataStreamA = new IndexAbstraction.DataStream(mb.dataStream("a"));
+        IndexAbstraction concreteIndexA = new IndexAbstraction.ConcreteIndex(
+            IndexMetadata.builder("a")
+                .settings(
+                    Settings.builder()
+                        .put(SETTING_NUMBER_OF_SHARDS, 1)
+                        .put(SETTING_NUMBER_OF_REPLICAS, 0)
+                        .put(SETTING_INDEX_VERSION_CREATED.getKey(), Version.CURRENT)
+                )
+                .build()
+        );
+        IndicesPermission.IsResourceAuthorizedPredicate predicate = predicate1.and(predicate2);
+        assertThat(predicate.test(dataStreamA), is(false));
+        assertThat(predicate.test(concreteIndexA), is(true));
```

Comment:
Nit: Can we add a few more assertions here for the more common cases of the `and` usage? e.g.:
1. resource is allowed by both the predicates with just the resourceNameMatcher
2. resource is allowed by the original predicate but *not* allowed by the limited predicate
3. The opposite of 2
4. Resource is denied by both (This is already covered by the test of `dataStreamA`)

Reply:
pushed [b8cd97e](https://github.com/elastic/elasticsearch/pull/92625/commits/b8cd97e51294304dcec8e0734e4adf209d9f20e8)